GreedyParameterPointSampler.C

```diff
@@ -0,0 +1,1420 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2013-2021, Lawrence Livermore National Security, LLC
+ * and other libROM project developers. See the top-level COPYRIGHT
+ * file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ *
+ *****************************************************************************/
+
+// Description: This class greedily selects parameter points
+//              for the construction of a ROM database.
+
+#include "GreedyParameterPointSampler.h"
+#include "HDFDatabase.h"
+#include "mpi.h"
+#include <cmath>
+#include <algorithm>
+#include <limits.h>
+#include <fstream>
+
+namespace CAROM {
+
+struct GreedyErrorIndicatorPoint
+createGreedyErrorIndicatorPoint(Vector* point, Vector* localROM)
+{
+    struct GreedyErrorIndicatorPoint result;
+    result.point = std::shared_ptr<Vector>(point);
+    result.localROM = std::shared_ptr<Vector>(localROM);
+    return result;
+}
+
+struct GreedyErrorIndicatorPoint
+createGreedyErrorIndicatorPoint(Vector* point, std::shared_ptr<Vector>& localROM)
+{
+    struct GreedyErrorIndicatorPoint result;
+    result.point = std::shared_ptr<Vector>(point);
+    result.localROM = localROM;
+    return result;
+}
+
+Vector
+getNearestPoint(std::vector<Vector> param_points, Vector point)
+{
+
+    int closest_point_index = getNearestPointIndex(param_points, point);
+    return param_points[closest_point_index];
+}
+
+int
+getNearestPointIndex(std::vector<Vector> param_points, Vector point)
+{
+
+    double closest_dist_to_points = INT_MAX;
+    int closest_point_index = -1;
+
+    for (int i = 0; i < param_points.size(); i++)
+    {
+        Vector diff;
```

Comment:
If there are many points, this could be optimized by some geometric search algorithm (e.g. maintaining a data structure that decomposes the domain into boxes and keeps the points in those boxes), instead of this exhaustive O(n) search. However, the greatest cost seems to be associated with evaluating ROMs at each parameter point? Or are ROMs only evaluated at a small subset of points? It is not clear to me yet whether such an optimization of this search would help in the end, considering other costs of the overall algorithm. In any case, it is not necessary in this PR, but something to think about maybe for the future.

Reply:
Sounds good. By the way, I think the idea I am referring to is called "spatial hashing."